client/handle_http.go

```diff
@@ -2122,31 +2126,52 @@ func runPut(request *http.Request, responseChan chan<- *http.Response, errorChan
 }
 
 // Walk a remote directory in a WebDAV server, emitting the files discovered
-func (te *TransferEngine) walkDirDownload(job *clientTransferJob, transfers []transferAttemptDetails, files chan *clientTransferFile, url *url.URL) error {
-	// Create the client to walk the filesystem
-	rootUrl := *url
-	if job.job.namespace.DirListHost != "" {
-		// Parse the dir list host
-		dirListURL, err := url.Parse(job.job.namespace.DirListHost)
+func (te *TransferEngine) walkDirDownload(job *clientTransferJob, transfers []transferAttemptDetails, files chan *clientTransferFile, remoteURL *url.URL) error {
+	// Query the director a PROPFIND to see if we can get our directory listing
+	resp, err := queryDirector("PROPFIND", remoteURL.Path, job.job.directorUrl+"/api/v1.0/director/origin")
+	if err != nil {
+		return err
+	}
+
+	collectionsUrl := *remoteURL
+	// If we get a "method not allowed", we go back to the old method
+	if resp.StatusCode == http.StatusMethodNotAllowed {
+		log.Debugln("PROPFIND against director failed, trying to get collections url from namespace")
+		// Check for a dir list host in namespace
+		if job.job.namespace.DirListHost != "" {
+			dirListURL, err := url.Parse(job.job.namespace.DirListHost)
+			if err != nil {
+				log.Errorln("Failed to parse dirlisthost from namespaces into URL:", err)
+				return err
+			}
+			collectionsUrl = *dirListURL
+		} else {
+			// Both methods to get directory listings failed
+			log.Errorln("Host for directory listings is unknown")
+			return errors.New("Host for directory listings is unknown")
```

Comment:
Does this result in duplicate error messages if we both log and return the same message?